internal/server/server.go

```diff
@@ -32,32 +29,43 @@ func Run(opts ...Option) error {
 		cfg.Logger = hclog.L()
 	}
 
-	// Setup our run group since we're going to be starting multiple
-	// goroutines for all the servers that we want to live/die as a group.
-	var group run.Group
-
-	// We first add an actor that just returns when the context ends. This
-	// will trigger the rest of the group to end since a group will not exit
-	// until any of its actors exit.
-	ctx, cancelCtx := context.WithCancel(cfg.Context)
-	cfg.Context = ctx
-	group.Add(func() error {
-		<-ctx.Done()
-		return ctx.Err()
-	}, func(error) { cancelCtx() })
-
-	// Setup our gRPC server.
-	if err := grpcInit(&group, &cfg); err != nil {
+	grpcServer, err := newGrpcServer(&cfg)
+	if err != nil {
 		return err
 	}
 
-	// Setup our HTTP server.
-	if err := httpInit(&group, &cfg); err != nil {
-		return err
+	errch := make(chan error)
+	go func() {
+		if err := grpcServer.start(); err != nil {
+			errch <- err
+		}
+	}()
+
+	httpEnabled := cfg.HTTPListener != nil
+	var httpServer *httpServer
+	if httpEnabled {
+		httpServer = newHttpServer(grpcServer.server, &cfg)
+		go func() {
+			if err := httpServer.start(); err != nil {
+				errch <- err
+			}
+		}()
 	}
 
-	// Run!
-	return group.Run()
+	ctx, cancel := context.WithCancel(cfg.Context)
+	defer cancel()
+
+	select {
+	case err := <-errch:
```

Comment:
Do we need to close here as well? i.e. if there is an error we should probably clean up the running goroutines?

Reply:
We also don't know which goroutine this error came from, so we need to make sure calling `close()` on an already closed thing is OK. I looked briefly and it seems so, but depending on the nature of the error maybe the http/grpc server isn't even around anymore?

Reply:
👍 good points. I'm now tracking which server exited, and shutting down the _other_ server. From my reading, it looks like when the server run processes exit they close their listeners, so you don't have to close them.

I'm also no waiting for each server routine to finish before exiting this function. This makes sure we have time to log the final error message before we quit.